File: packages/go/core/store/buffer.go
package store

import (
	"sync"
	"time"
)

type AsyncWriteBuffer struct {
	backend       HitStore
	queue         []HitEvent
	mu            sync.Mutex
	flushInterval time.Duration
	maxBuffer     int
	batchSize     int
	done          chan struct{}
	wg            sync.WaitGroup
}

func NewAsyncWriteBuffer(backend HitStore, flushInterval time.Duration, maxBuffer, batchSize int) *AsyncWriteBuffer {
	if flushInterval == 0 {
		flushInterval = 2 * time.Second
	}
	if maxBuffer == 0 {
		maxBuffer = 10_000
	}
	if batchSize == 0 {
		batchSize = 100
	}
	return &AsyncWriteBuffer{
		backend:       backend,
		flushInterval: flushInterval,
		maxBuffer:     maxBuffer,
		batchSize:     batchSize,
		done:          make(chan struct{}),
	}
}

func (b *AsyncWriteBuffer) Start() {
	b.wg.Add(1)
	go func() {
		defer b.wg.Done()
		b.flushWorker()
	}()
}

func (b *AsyncWriteBuffer) Record(hit HitEvent) error {
	b.mu.Lock()
	defer b.mu.Unlock()
	if len(b.queue) >= b.maxBuffer {
		b.queue = b.queue[1:] // drop oldest on overflow
	}
	b.queue = append(b.queue, hit)
	return nil
}

func (b *AsyncWriteBuffer) flushWorker() {
	ticker := time.NewTicker(b.flushInterval)
	defer ticker.Stop()
	for {
		select {
		case <-ticker.C:
			b.flush()
		case <-b.done:
			b.flush()
			return
		}
	}
}

func (b *AsyncWriteBuffer) flush() {
	b.mu.Lock()
	if len(b.queue) == 0 {
		b.mu.Unlock()
		return
	}
	n := len(b.queue)
	if n > b.batchSize {
		n = b.batchSize
	}
	batch := make([]HitEvent, n)
	copy(batch, b.queue[:n])
	b.queue = b.queue[n:]
	b.mu.Unlock()

	for _, hit := range batch {
		b.backend.Record(hit) //nolint:errcheck
	}
}

// FlushAll drains the queue synchronously. Used in tests.
func (b *AsyncWriteBuffer) FlushAll() {
	for {
		b.mu.Lock()
		n := len(b.queue)
		b.mu.Unlock()
		if n == 0 {
			break
		}
		b.flush()
	}
}

func (b *AsyncWriteBuffer) RecentHits(query HitQuery) ([]HitEvent, error) {
	return b.backend.RecentHits(query)
}

func (b *AsyncWriteBuffer) EndpointSummaries(sinceDays int) ([]EndpointSummary, error) {
	return b.backend.EndpointSummaries(sinceDays)
}

func (b *AsyncWriteBuffer) TotalSummary(sinceDays int) (TotalSummary, error) {
	return b.backend.TotalSummary(sinceDays)
}

func (b *AsyncWriteBuffer) Close() error {
	close(b.done)
	b.wg.Wait()
	return b.backend.Close()
}

File: packages/javascript/src/core/store/buffer.ts
import { HitStore, HitEvent, HitQuery, EndpointSummary, TotalSummary } from './base'

export class AsyncWriteBuffer extends HitStore {
  private queue: HitEvent[] = []
  private timer: ReturnType<typeof setInterval> | null = null

  constructor(
    private readonly backend: HitStore,
    private readonly flushInterval = 2000,
    private readonly maxBuffer = 10_000,
    private readonly batchSize = 100,
  ) {
    super()
  }

  start(): void {
    this.timer = setInterval(() => { void this._flush() }, this.flushInterval)
    // Don't keep the Node.js process alive just for the flush timer
    this.timer.unref()
  }

  async record(hit: HitEvent): Promise<void> {
    if (this.queue.length >= this.maxBuffer) {
      this.queue.shift() // drop oldest on overflow
    }
    this.queue.push(hit)
  }

  private async _flush(): Promise<void> {
    const batch = this.queue.splice(0, this.batchSize)
    for (const hit of batch) {
      await this.backend.record(hit)
    }
  }

  /** Flush all pending hits synchronously — useful in tests and graceful shutdown. */
  async flushAll(): Promise<void> {
    while (this.queue.length > 0) {
      await this._flush()
    }
  }

  async recentHits(query?: HitQuery): Promise<HitEvent[]> {
    return this.backend.recentHits(query)
  }

  async endpointSummaries(since_days?: number): Promise<EndpointSummary[]> {
    return this.backend.endpointSummaries(since_days)
  }

  async totalSummary(since_days?: number): Promise<TotalSummary> {
    return this.backend.totalSummary(since_days)
  }

  async close(): Promise<void> {
    if (this.timer) clearInterval(this.timer)
    await this.flushAll()
    await this.backend.close()
  }
}
